src/goto-instrument/contracts/contracts.cpp

```diff
@@ -311,137 +358,154 @@ void code_contractst::check_apply_loop_contracts(
 
-  insert_before_swap_and_advance(goto_function.body, loop_head, generated_code);
+  // Insert the second block of pre_loop_head_instrs: the havocing code.
+  // We do not `add_pragma_disable_assigns_check`,
+  // so that the enclosing scope's assigns clause instrumentation
+  // would pick these havocs up for inclusion (subset) checks.
+  goto_function.body.destructive_insert(loop_head, pre_loop_head_instrs);
 
   // Generate: assume(invariant) just after havocing
-  // We use a block scope to create a temporary assumption,
-  // and immediately convert it to goto instructions.
+  // We use a block scope for assumption, since it's immediately goto converted,
+  // and doesn't need to be kept around.
   {
-    code_assumet assumption{invariant_expr()};
-    assumption.add_source_location() = loop_head->source_location();
-    converter.goto_convert(assumption, generated_code, mode);
+    code_assumet assumption{invariant};
+    assumption.add_source_location() = loop_head_location;
+    converter.goto_convert(assumption, pre_loop_head_instrs, mode);
   }
 
   // Create fresh temporary variables that store the multidimensional
   // decreases clause's value before and after the loop
   for(const auto &clause : decreases_clause.operands())
   {
     const auto old_decreases_var =
-      new_tmp_symbol(
-        clause.type(), loop_head->source_location(), mode, symbol_table)
+      new_tmp_symbol(clause.type(), loop_head_location, mode, symbol_table)
         .symbol_expr();
-    generated_code.add(goto_programt::make_decl(
-      old_decreases_var, loop_head->source_location()));
+    pre_loop_head_instrs.add(
+      goto_programt::make_decl(old_decreases_var, loop_head_location));
     old_decreases_vars.push_back(old_decreases_var);
 
     const auto new_decreases_var =
-      new_tmp_symbol(
-        clause.type(), loop_head->source_location(), mode, symbol_table)
+      new_tmp_symbol(clause.type(), loop_head_location, mode, symbol_table)
         .symbol_expr();
-    generated_code.add(goto_programt::make_decl(
-      new_decreases_var, loop_head->source_location()));
+    pre_loop_head_instrs.add(
+      goto_programt::make_decl(new_decreases_var, loop_head_location));
     new_decreases_vars.push_back(new_decreases_var);
   }
 
   // TODO: Fix loop contract handling for do/while loops.
   if(loop_end->is_goto() && !loop_end->get_condition().is_true())
   {
     log.error() << "Loop contracts are unsupported on do/while loops: "
-                << loop_head->source_location() << messaget::eom;
+                << loop_head_location << messaget::eom;
     throw 0;
 
     // non-deterministically skip the loop if it is a do-while loop.
-    generated_code.add(goto_programt::make_goto(
-      loop_end,
-      side_effect_expr_nondett(bool_typet(), loop_head->source_location())));
+    pre_loop_head_instrs.add(goto_programt::make_goto(
+      loop_end, side_effect_expr_nondett(bool_typet(), loop_head_location)));
```

Comment:
What does this jump model exactly ? Is this the sound thing to do ?

Reply:
It's right after a `throw 0`, so ignored for now. This needs to be fixed for handling `do ... while` loop contracts, but the semantics of loop invariants on those loops is still under discussion (whether the firs iteration should be "pre" code or "loop" code).